sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,602 @@
+      device_config |= 0x1 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  // Ensure that we actually have a uint12_t.
+  if (config.rx_fifo_timeout > 0xfff) {
+    return kDifSpiDeviceResultBadArg;
+  }
+  device_config |= (config.rx_fifo_timeout & SPI_DEVICE_CFG_TIMER_V_MASK)
+                   << SPI_DEVICE_CFG_TIMER_V_OFFSET;
+
+  uint16_t rx_fifo_start = 0x0;
+  uint16_t rx_fifo_end = config.rx_fifo_len - 1;
+  uint16_t tx_fifo_start = config.rx_fifo_len;
+  uint16_t tx_fifo_end = config.rx_fifo_len + config.tx_fifo_len - 1;
+  if (tx_fifo_end >= kDifSpiDeviceSramLen) {
+    // We've overflown the SRAM region...
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t rx_fifo_bounds = 0;
+  rx_fifo_bounds |= (rx_fifo_start & SPI_DEVICE_RXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_RXF_ADDR_BASE_OFFSET;
+  rx_fifo_bounds |= (rx_fifo_end & SPI_DEVICE_RXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_RXF_ADDR_LIMIT_OFFSET;
+
+  uint32_t tx_fifo_bounds = 0;
+  tx_fifo_bounds |= (tx_fifo_start & SPI_DEVICE_TXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_TXF_ADDR_BASE_OFFSET;
+  tx_fifo_bounds |= (tx_fifo_end & SPI_DEVICE_TXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_TXF_ADDR_LIMIT_OFFSET;
+
+  spi->rx_fifo_base = rx_fifo_start;
+  spi->rx_fifo_len = config.rx_fifo_len;
+  spi->tx_fifo_base = tx_fifo_start;
+  spi->tx_fifo_len = config.tx_fifo_len;
+
+  // Now that we know that all function arguments are valid, we can abort
+  // all current transactions and begin reconfiguring the device.
+  dif_spi_device_result_t abort_err = dif_spi_device_abort(spi);
+  if (abort_err != kDifSpiDeviceResultOk) {
+    return abort_err;
+  }
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_CFG_REG_OFFSET, device_config);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_RXF_ADDR_REG_OFFSET,
+                      rx_fifo_bounds);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_TXF_ADDR_REG_OFFSET,
+                      tx_fifo_bounds);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  // Set the |abort| bit, and then spin until |abort_done| is asserted.
+  mmio_region_nonatomic_set_bit32(spi->base_addr, SPI_DEVICE_CONTROL_REG_OFFSET,
+                                  SPI_DEVICE_CONTROL_ABORT);
+  while (!mmio_region_get_bit32(spi->base_addr, SPI_DEVICE_STATUS_REG_OFFSET,
+                                SPI_DEVICE_STATUS_ABORT_DONE)) {
+  }
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_get(const dif_spi_device_t *spi,
+                                               dif_spi_device_irq_type_t type,
+                                               bool *flag_out) {
+  if (spi == NULL || flag_out == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t bit_index = 1;
+  switch (type) {
```

Comment:
Split this switch statement into a separate function, it is used across multiple functions.

Reply:
It's not. The values assigned in each switch are actually different macros.

Reply:
True, I think I have issue open regarding this - I think although those all are separate macros, they always will have the same values, as they are extrapolated from the same #1206 .

Reply:
If they wind up getting unified, please let me know so that I can unify these switches. I had done the work earlier but had to undo it because the macros are different...

Reply:
Will, do. There wasn't much interest in that issue, so I guess it probably won't and reggen is not the most intuitive tool on this project, so I avoid touching it as much as I can. :)